app/Console/Commands/WaitingLists/CheckMembersMeetActivityRules.php

```diff
@@ -0,0 +1,160 @@
+            $this->processWaitingListAccountsMarkedForRemoval($waitingList);
+       }
+       
+    }
+
+    /**
+     * Retrieves training waiting lists which enforce hour requirements
+     */
+    protected function retrieveHourEnforcedWaitingLists()
+    {
+        return WaitingList::where('enforce_hour_requirement', '=', 'yes')->get();
+    }
+
+    /**
+     * Identifies accounts on enforced waiting lists who have fallen below the hour requirement
+     * 
+     * @param WaitingList $waitingList
+     */
+    protected function identifyNewNonEligibleWaitingListAccounts(WaitingList $waitingList)
+    {
+        // We only want to find Active waiting list accounts which are not already marked for removal
+        $waitingListAccounts                                        = $waitingList->accounts
+            ->filter(function($account){
+                return (
+                    $account->pivot->current_status->name == 'Active' &&
+                    (
+                        is_null($account->pivot->pending_removal->removal_date ?? null) ||
+                        $account->pivot->pending_removal->status != 'Pending'
+                    )
+                );
+            });
+        
+        foreach($waitingListAccounts as $account){
+            // Calculate the total activity within the last 3 months
+            if(isset($this->accountActivityCache[$account->id])){
+                $activeMinutes                                      = $this->accountActivityCache[$account->id];
+            } else {
+                $activeMinutes                                      = Atc::where('account_id', $account->id)
+                    ->whereDate('disconnected_at', '>=', Carbon::parse('3 months ago'))
+                    ->isUk()
+                    ->sum('minutes_online');
+                $this->accountActivityCache[$account->id]           = $activeMinutes;
+            }
+
+            // When the hours are less than the requirement, set a removal date and notify the member
+            if(((int)$activeMinutes ?? 0) < $this->minutesRequired){
+                $removalDate                                        = Carbon::now();
+                $removalDate->addDays(31); // When calculating date diff, this will be a difference of 30 days
+                $account->pivot->addPendingRemoval($removalDate);
+                $account->notify(new WaitingListRemovalAdded($waitingList->name, $removalDate));
+            }
+        }
+    }
+
+
+    /**
+     * Performs two actions:
+     * 
+     * - Checks to see if members who are pending removal have now met the hour criteria
+     * - Sends reminder emails as per policy schedule, and actions waiting list removal
+     * 
+     * @param WaitingList $waitingList
+     */
+    protected function processWaitingListAccountsMarkedForRemoval(WaitingList $waitingList)
+    {
+         $waitingListAccounts                                        = $waitingList->accounts
```

Comment:
This is what I mean - here, the logic is duplicated, where you could have just passed in a list of accounts you want to start the removal process for.

Reply:
Hi @AxonC 

This function accepts a waiting list, and then discovers accounts which are marked for removal for that waiting list.

I don't fully understand how the logic is duplicated, as each time the function is called, the waiting list will be different - so line 113 retrieves the relevant accounts for the current waiting list?